### PR TITLE
To container throw on missing

### LINKED ISSUE #501: Feature: throw_on_missing keyword argument for OmegaConf.to_container
# Feature:
Currently, calling `OmegaConf.to_container(cfg)` can result in a python object that contains some `"???"` MISSING strings.

I propose to add a `throw_on_missing` keyword argument to `OmegaConf.to_container`, so that the structure returned by `OmegaConf.to_container(cfg, throw_on_missing=True)` will raise `omegaconf.MissingMandatoryValue` if any missing values are encountered.

This way, if no `MissingMandatorValue` is raised, clients can be assured that the result of `to_container` does not contain any `"???"` strings.

To maintain backward compatibility, the default would be `throw_on_missing=False`.

### PR DESCRIPTION
**Pull request by Jasha10**
Closes #501

### CODE REVIEW

**Comment by Jasha10** [comment]
Still needs test coverage for the new `throw_on_missing_flag`.

**Review by omry — COMMENTED** [review]
It will probably be simpler to 
node = conf._get_node(key, throw_on_missing=True)

Try to line the error message with the normal error for missing values (`raise MissingMandatoryValue("Missing mandatory value: $FULL_KEY")`.

**Comment by Jasha10** [comment]
Rebased onto master.

**Comment by Jasha10** [comment]
> It will probably be simpler to
> node = conf._get_node(key, throw_on_missing=True)

`conf._get_node` does not have a `throw_on_missing` keyword option.
Do you mean that I should add a `throw_on_missing` keyword argument to the `_get_node` signature?

**Comment by omry** [comment]
> Do you mean that I should add a throw_on_missing keyword argument to the _get_node signature?

This is what I had in mind, yes. I added it in #513.

File: tests/test_errors.py
Comment (Jasha10) [reply tests/test_errors.py:1074]: > They should probably be moved to test_omegaconf.py, or maybe even to a dedicated test_to_container.py file.

Ok, I'll move the `to_container` tests into a `test_to_container.py` file.

> As a side note:
> test_errors should probably be reorganized and broken into smaller functions (while reusing the error validation logic).

I'll open a draft PR for refactoring `test_errors.py`

File: omegaconf/basecontainer.py
Comment (omry) [inline omegaconf/basecontainer.py:220]: can you refactor out a small local raise_missing_value_exception() function to reduce the amount of new code?

File: omegaconf/omegaconf.py
Comment (omry) [inline omegaconf/omegaconf.py:514]: Side note:
Currently we do not generate API docs.
It would actually be useful as the API surface area increases.

**Review by omry — CHANGES_REQUESTED** [review]
Marking as request-changes (see prev comments)

**Comment by Jasha10** [comment]
Rebased onto `master`

File: omegaconf/basecontainer.py
Comment (Jasha10) [reply omegaconf/basecontainer.py]: Done, in https://github.com/omry/omegaconf/pull/503/commits/7b49aea8781c97a1515a7f723f78a141054e4dfe.

File: omegaconf/omegaconf.py
Comment (Jasha10) [reply omegaconf/omegaconf.py:515]: Created issue #536 to keep track of this idea.

File: tests/test_errors.py
Comment (Jasha10) [reply tests/test_errors.py:1088]: > They should probably be moved to test_omegaconf.py, or maybe even to a dedicated test_to_container.py file.

I've submitted PR #538 to move `to_container` tests into a new `test_to_container.py` file.

**Review by omry — COMMENTED** [review]
Can't comment on code you didn't change but I found what I was thinking about with the _get_node() comment:

```python
            retdict: Dict[str, Any] = {}
            for key in conf.keys():	            for key in conf.keys():
                node = conf._get_node(key)	                node = conf._get_node(key)
```

In above _get_node(), what happens if you use `conf._get_node(key, throw_on_missing=throw_on_missing)`?

(You will probably need to do the same for ListConfig somwehere).

**Comment by Jasha10** [comment]
> In above _get_node(), what happens if you use `conf._get_node(key, throw_on_missing=throw_on_missing)`?

That works! There are a few test failures in `test_errors.py` which look like this:
```
AssertionError: Regex pattern 'Missing mandatory value: subcfg.x' does not match 'Missing mandatory value'.
```
If it's ok with you to change the error message, I'll make the change to using `conf._get_node(key, throw_on_missing_value=throw_on_missing)` and will update the `test_errors.py` error messages.

File: omegaconf/dictconfig.py
Comment (Jasha10) [inline omegaconf/dictconfig.py:454]: I had to change `DictConfig._get_node` (and a similar change for `ListConfig._get_node` so that the `MissingMandatoryValue` will have more information about what `key` and what `node` caused the error.

File: omegaconf/listconfig.py
Comment (omry) [inline omegaconf/listconfig.py:407]: Nice, but add corresponding tests of _get_node that confirms that this is working (especially the slice case).
Feel free to improve test_get_node_throw_on_missing_value to do that instead of making test_errors even bigger.

**Review by omry — CHANGES_REQUESTED** [review]
much cleaner.

File: omegaconf/listconfig.py
Comment (Jasha10) [reply omegaconf/listconfig.py:407]: By adding tests (https://github.com/omry/omegaconf/pull/503/commits/7d7026fb8cc8463e0a5c291abdca6a6549176a25) I discovered that there was an error in the slice case. I have seen the light. Test everything.

File: omegaconf/listconfig.py
Comment (omry) [inline omegaconf/listconfig.py:400]: This is a lot of custom logic for pretty low return.
I can live with a simpler handling for slice errors with missing (no need to report the exact first missing key).
I think just use the slice as is without reporting the exact index. it would also simplify your tests.

**Review by omry — CHANGES_REQUESTED** [review]

**Comment by Jasha10** [comment]
Currently code coverage tests are failing because the `except` blocks from the latest commit https://github.com/omry/omegaconf/pull/503/commits/fe32082fca206722323c2387f04d2074108c1297 have no coverage. Those try/except statements were added in anticipation of #545, and I anticipate that the coverage issue will be resolved one this branch is rebased onto that one.

File: omegaconf/listconfig.py
Comment (Jasha10) [reply omegaconf/listconfig.py]: Done in https://github.com/omry/omegaconf/pull/503/commits/01d53955351a9db389cf7ac3ee42b3f0085c081e

File: omegaconf/basecontainer.py
Comment (omry) [inline omegaconf/basecontainer.py:238]: 1. Why is _dereference_node throwing a MandatoryMissingValue is throw_on_missing is False?
2. Why are you passing False to throw_on_missing and not the input flag?

Comment (Jasha10) [reply omegaconf/basecontainer.py:238]: > 1. Why is _dereference_node throwing a MandatoryMissingValue is throw_on_missing is False?

This is because we are calling `_dereference_node` with `throw_on_resolution_failure=True`.
If #545 is merged in it's current state, `_dereference_node` will raise `MandatoryMissingValue` if `throw_on_resolution_failure==True` and the interpolation cannot be resolved.

This being said, I actually think it makes more sense to change this from
```python
_dereference_node(..., throw_on_resolution_failure=True)
```
to
```python
_dereference_node(..., throw_on_resolution_failure=throw_on_missing)
```
so that resolution failure will not cause error if the user does not specify the `throw_on_missing` option. I'll add tests to this effect...

Comment (Jasha10) [reply omegaconf/basecontainer.py:238]: > 2. Why are you passing False to throw_on_missing and not the input flag?

The current workflow is:
- first, get the node and (optionally) throw if missing:
```python
node = conf._get_node(key, throw_on_missing_value=throw_on_missing)
```
- second: if `resolve==True`, dereference the node:
```python
node = node._dereference_node(throw_on_missing=False, throw_on_resolution_failure=True)
```
Since the `_get_node` call already has `throw_on_missing_value=throw_on_missing`, it would be redundant to use the `throw_on_missing` flag in the `_dereference_node` call as well.

**Comment by Jasha10** [comment]
#545 will change the behavior of `_dereference_node` significantly enough that it makes sense to put this PR on hold until #545 gets merged.

**Comment by Jasha10** [comment]
Marking this PR as needing possible API redesign after #502 gets merged.

**Comment by Jasha10** [comment]
This PR is old; closing in favor of #730.

**Pull request closed without merging**
